Bugfix/various cppcheck improvements

rmelotte opened the pull request:
A couple of improvements to the cppcheck-diff job that I worked on in the background.

- output issues with relative paths, this makes it possible to generate and use the list of issues locally. 
- add a shell wrapper for cppcheck-new-issues.sh to run the cppcheck-diff job locally.
- report the solved issues, and update the list of issues automatically. This way, people just have to run the script and solved issues are removed from the list automatically (or it can also be copied from gitlab).
- fix some issues in tlvf.py
- ignore the tlvf folder, since we can't fix them all at the moment. This avoids the cppcheck-diff job from bugging people with issues in tlvf.py we can't solve easily.

rmelotte committed:
tlvf: use a reinterpret_cast to handle param in init and m_type

tlvf.py was using a C-style cast in init, but it looks like the
intention is to do a reinterpret_cast.

Since we want to move away from C-style casts, replace it with
reinterpret_cast.

Also do the same for m_type.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
ci: cppcheck-new-issues.sh: remove the additional printed line

When a new issue was found, the first line of the next issue was also
printed.

The -A option was used to display the caret that points the location
of the issue in the code.
However, it turns out the position of the caret is anyway incorrect
because the code is not on a newline.

When we generate the list of issues, remove the carets lines.
Also remove the -A option when printing the issues.

Also remove the unneeded single quote surrounding the output. Because
of this change, we have to update the whole list of existing issues.

It also requires an update to the prplmesh-builder:alpine-3.11.3
image.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
ci: cppcheck-new-issues.sh: report solved issues

When people solve new issues over time, they need to be deleted from
the list of existing issues.

Generate a new "cppcheck_existing_issues.txt" file by removing solved
issues, and save it as an artifact.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
tools: cppcheck: output file names relative to rootdir

The issues in ci/cppcheck/cppcheck_existing_issues.txt contain
absolute paths (for some of them).

This is annoying because it prevents people to run
cppcheck-new-issues.sh locally.

Use the "rp" option so that file paths output by cppcheck are always
relative to the top-level folder.

Even with the "rp" option, some issues still contain absolute paths:
- [noValidConfiguration]
- [ConfigurationNotChecked]

To circumvent this, pipe cppcheck's output to sed to filter out those
paths. This require an update of the following image:
prplfoundationinc/prplmesh-builder:alpine-3.11.3

sed's "-u" option is used to display issues as they come instead of
only displaying them when sed's buffer is full.

The list of existing cppcheck issues is also updated to drop the
absolute paths.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
cppcheck: move the wrapper to tools/docker

Since cppcheck.sh won't be the only wrapper that runs with docker,
it's best to move it. clang-format wrapper is already under
tools/docker, so it makes sense to also put it there.

While we're at it, add a license header.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
tools: docker: cppcheck: run with the same user id as the on the host

If the user id is not specified, the docker container runs as root.

Provide the user id so that docker doesn't create files owned by root
in the local tree. If the container is run with sudo, SUDO_UID is used
to use the original user id.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
tools: docker: also run cppcheck-new-issues.sh

Multiple developers like to be able to run locally the jobs that we
run in CI.

Because cppcheck results are so dependent on the version being used,
it needs to be run with the same docker image that is used in CI.

When running tools/docker/cppcheck.sh, also run what the
"cppcheck-diff" job runs, in the same docker image.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
tools: cppcheck-new-issues.sh: ignore tlvf

Every time someone adds or modifies a tlv, cppcheck-diff will report
new issues because tlvf.py generate them.

Since we cannot easily fix them all at once, remove them from the
checks and open a new task to fix them:
https://github.com/prplfoundation/prplMesh/issues/1191

Also update the list of known cppcheck issues to remove the issues
regarding tlvf, since the cppcheck-diff job will report they are now
solved.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
framework: tlvf.py: add missing override specifier

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
tools: cppcheck: sort the issues alphabetically

This is a preparatory commit. Having the issues sorted makes it easier
to later compare them.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
tlvf: update autogenerated files

Also remove solved cppcheck issues.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

arnout commented on framework/tlvf/AutoGenerated/src/tlvf/test/tlvVarList.cpp:
```diff
@@ -564,7 +564,7 @@ bool tlvTestVarList::init()
         complex_list->class_swap();
     }
-    m_var1 = (cInner*)m_buff_ptr__;
+    m_var1 = reinterpret_cast<cInner*>(m_buff_ptr__);
```
This ended up in the wrong commit.

arnout commented on ci/cppcheck/cppcheck-new-issues.sh:
```diff
@@ -25,8 +25,19 @@ do
+
+mv "$rootdir"/cppcheck_existing_issues.txt "$rootdir"/ci/cppcheck/cppcheck_existing_issues.txt
+
+if ! git diff -U0 --exit-code -- "$rootdir"/ci/cppcheck/cppcheck_existing_issues.txt ; then
```
I think this should only be done if status was still 0. Otherwise, you'll get both the report of the new issue above, and again in the diff here.

arnout commented on ci/cppcheck/cppcheck-new-issues.sh:
```diff
@@ -25,8 +25,19 @@ do
 done
 
+# Generate a new list of existing issues by removing the ones that are now solved
+comm -12 <(sort "$rootdir"/cppcheck_results.txt) <(sort "$rootdir"/ci/cppcheck/cppcheck_existing_issues.txt)  > "$rootdir"/cppcheck_existing_issues.txt
```
This relies very heavily on the order of the cppcheck_results.txt file to be stable. We run cppcheck with just the rootdir argument. I don't know how it iterates, but I assume it uses directory order without sorting. In that case, the order is not stable, so we may get different results on different runs.

rmelotte replied:
The list of issues is always sorted first, so it should always produce the same result. Or am I misunderstanding something?

arnout commented on ci/cppcheck/cppcheck-new-issues.sh:
```diff
@@ -25,8 +25,19 @@ do
+# Generate a new list of existing issues by removing the ones that are now solved
+comm -12 <(sort "$rootdir"/cppcheck_results.txt) <(sort "$rootdir"/ci/cppcheck/cppcheck_existing_issues.txt)  > "$rootdir"/cppcheck_existing_issues.txt
+
+mv "$rootdir"/cppcheck_existing_issues.txt "$rootdir"/ci/cppcheck/cppcheck_existing_issues.txt
```
Why move it instead of directly generating it in the right location?

Actually, why move it on top of the existing one and use git diff? You could just as well leave it in the root directory and use normal diff. And if you want to avoid needing to install normal diff, you can use git diff with two files as arguments even if one of them is not tracked.

arnout commented on ci/cppcheck/cppcheck_existing_issues.txt:
```diff
@@ -1,399 +1,762 @@
-'/builds/prpl-foundation/prplMesh/framework/tlvf/AutoGenerated/src/tlvf/wfa_map/tlvSteeringPolicy.cpp: style: The if condition is the same as the previous if condition [duplicateCondition]    if (m_parse__) {
-'/builds/prpl-foundation/prplMesh/framework/tlvf/AutoGenerated/src/tlvf/wfa_map/tlvSteeringPolicy.cpp: style: C-style pointer casting [cstyleCast]    m_radio_ap_control_policy_list = (sRadioApControlPolicy*)m_buff_ptr__;
+'common/beerocks/bcl/include/bcl/network/socket.h: style: Statements following return, break, continue, goto or throw will never be executed. [unreachableCode]        m_error.clear();
+        ^'
```
There seems to be some commit reordering going on here, because the carets are re-introduced here...

arnout commented on tools/docker/cppcheck-diff.sh:
```diff
@@ -0,0 +1,12 @@
+scriptdir="$(cd "${0%/*}" && pwd)"
+rootdir="$(realpath "$scriptdir/../..")"
+
+docker run --rm -it -w "$rootdir" -v "$rootdir:$rootdir" --user="${SUDO_UID:-$UID}" prplfoundationinc/prplmesh-builder:alpine-3.11.3 "$rootdir"/ci/cppcheck/cppcheck-new-issues.sh .
```
(nitpick) I'd prefer this as an option to cppcheck.sh instead, to avoid a proliferation of small wrapper scripts...

rmelotte replied:
Done

arnout commented on tools/docker/cppcheck-diff.sh:
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+###############################################################
+# SPDX-License-Identifier: BSD-2-Clause-Patent
+# SPDX-FileCopyrightText: 2019-2020 the prplMesh contributors (see AUTHORS.md)
```
This file 2020 only (but becomes irrelevant with my comment below)

arnout commented on framework/tlvf/tlvf.py:
```diff
@@ -779,7 +779,7 @@ def addClassParam(self, obj_meta, param_name, param_type, param_type_info, param
                 else:
                     # add default value to init func
-                    lines_cpp.append("m_%s = (%s*)m_%s__;" % ( param_name, param_type, self.MEMBER_BUFF_PTR))
+                    lines_cpp.append("m_%s = reinterpret_cast<%s*>(m_%s__);" % ( param_name, param_type, self.MEMBER_BUFF_PTR))
```
Please squash this in the other commit doing the same thing somewhere else.

rmelotte replied:
Done

arnout commented on framework/tlvf/AutoGenerated/include/tlvf/test/tlvVarList.h:
```diff
@@ -48,7 +48,7 @@ class tlvTestVarList : public BaseClass
         std::shared_ptr<cInner> create_complex_list();
         bool add_complex_list(std::shared_ptr<cInner> ptr);
-        bool isPostInitSucceeded();
+        bool isPostInitSucceeded() override;
```
Again, this ended up in the wrong commit.

RanRegev replied:
also autogenerated tlvf are now ignored.

rmelotte replied:
Since the change to the autogenerated files was so small, I intentionally kept it in the same commit as the change to tlvf.py.

arnout commented on ci/cppcheck/cppcheck-new-issues.sh:
```diff
@@ -13,6 +13,10 @@ rootdir="${scriptdir%/*/*}"
+# Sort the list of issues to have a stable list, easier to compare:
+sort "$rootdir/cppcheck_results.txt" > "$rootdir/cppcheck_results_sorted.txt"
+mv "$rootdir/cppcheck_results_sorted.txt" "$rootdir/cppcheck_results.txt"
+
```
It would actually make more sense if cppcheck.sh would not create any file but simply output to stdout, and that it was the CI job that creates the output file. I find it pretty annoying that running a tool like that creates a file (which BTW isn't even in gitignore).

With that, you could use a nice pipeline here: `cppcheck.sh | sed | sort > results`.

RanRegev replied:
I agree with @arnout - scripts that generate files are less compose-able and are therefore harder to use in other contents. A mid solution is to add an _option_ to output to a file and use stdout as the default for this option.

arnout commented on .gitlab-ci.yml:
```diff
@@ -66,7 +66,7 @@ flake8-diff:
   script:
     - apk add --update --no-cache git
-    - git diff -U0 origin/master | flake8 --verbose --diff
+    -  git diff -U0 origin/master -- . ':(exclude)framework/tlvf/tlvf.py' | flake8 --verbose --diff
```
#1249 removes flake8-diff entirely instead, which IMO is better :-)

rmelotte replied:
Done

arnout reviewed (approved):
Lots of small comments, as usual. Nothing really stands out, so I'm fine with merging as is (except the last commit has to be dropped because it conflicts with #1249). We shouldn't spend too much time on this stuff.

MuteSpirit commented:
Looks like there are similar issues with C-style type casting in next lines of `framework/tlvf/tlvf.py`
* line 778
```
lines_h.append( "%sreturn (%s)(%s);" % (self.getIndentation(1), param_type, ret_const_val) )
```
* line 834
```
                    lines_cpp.append( "%sreturn (%s%s&)(*m_%s);" % (self.getIndentation(1), const, param_type, param_name) )  
```
* line 864
```
            lines_cpp.append("m_%s = (%s*)m_%s__;" % (param_name, param_type, self.MEMBER_BUFF_PTR) )
```
* line 1025
```
                lines_cpp.append( "%sreturn ((%s*)m_%s);" % (self.getIndentation(1), param_type, param_name) )
```

Were they missed or will be fixed another time?

RanRegev commented on framework/tlvf/tlvf.py:
```diff
@@ -745,7 +745,7 @@ def addClassParam(self, obj_meta, param_name, param_type, param_type_info, param
 
                 # Add param handling to init function
-                lines_cpp.append("m_%s = (%s*)m_%s__;" % ( param_name, param_type, self.MEMBER_BUFF_PTR))
+                lines_cpp.append("m_%s = reinterpret_cast<%s*>(m_%s__);" % ( param_name, param_type, self.MEMBER_BUFF_PTR))
```
C style cast and reinterpret cast when dealing with object pointers have no much different in the "high level" point of view: effectively you _guess_ what the bytes you point to means, and you tell the compiler to think the same as you.
I didn't follow the code but how can we be sure that `self.MEMBER_BUFF_PTR` is really of `param_type`. And if we are sure it is, why didn't we declare it to be such from the first place?

RanRegev commented on tools/docker/static-analysis/cppcheck.sh:
```diff
@@ -21,10 +21,11 @@ run_cppcheck() {
              -q \
              "${INCLUDES[@]}" \
-             "$@" 2>&1 | tee "$OUTPUT_FILE"
+             "$@" 2>&1 | sed "s|$rootdir/||g" | tee "$OUTPUT_FILE"
```
+1 for replacing the `/` command with another one because of having forward slash in the search.
I would use `@` but `|` is good too.

RanRegev commented on tools/docker/cppcheck.sh:
```diff
@@ -10,4 +10,4 @@
 rootdir="$(realpath "$scriptdir/../..")"
 
-docker run --rm -it -w "$rootdir" -v "$rootdir:$rootdir"  prplfoundationinc/prplmesh-builder:alpine-3.11.3 "$rootdir"/tools/docker/static-analysis/cppcheck.sh -j"$(nproc)" .
+docker run --rm -it -w "$rootdir" -v "$rootdir:$rootdir" --user="${SUDO_UID:-$UID}"  prplfoundationinc/prplmesh-builder:alpine-3.11.3 "$rootdir"/tools/docker/static-analysis/cppcheck.sh -j"$(nproc)" .
```
I feel uncomfortable with this change.
I didn't dive into the details, so it might be that everything is OK. However I would like the author (@rmelotte) to reference in the final decision the following:
* https://echorand.me/posts/docker-user-namespacing-remap-system-user/
* https://github.com/docker/docker-ce/blob/531930f3294c31db414f17f80fa8650d4ae66644/components/engine/api/types/container/host_config.go#L169
* and also other work that might have been done in this area by others in prplMesh

RanRegev reviewed (approved):
A. I miss documentation regarding the way we use cppcheck: CI, local, results, sort, diff, etc. I found [only few lines](https://github.com/prplfoundation/prplMesh/wiki/Testing-in-CI#cpp-check) in the wiki that explains how to run cppcheck locally. Maybe there is more documentation somewhere?
B. There are few commits in this PR that change things in tlvf.py that in turn change the AutoGenerated result. However, the entire AutoGenerated directory is ignored by another commit in this PR. So from the cppcheck point of view the changes are not checked, so maybe it is better to introduce these changes into another PR, unrelated to cppcheck improvements.

rmelotte commented:
I started solving some of the issues in tlvf.py, but then there were too many to solve in one go.
The reason TLVs are later ignored is to save other developers the time that they would have to spend dealing with those every time they add/modify a TLV.

rmelotte commented:
I addressed some of the comments, there are still some others valid and left open.
I would like to still merge it as is because:
- working on them is not a priority at moment.
- this PR already allows running the cppcheck-diff job locally and also makes it easier to add new tlvs (since issues coming from tlvf.py are ignored)
